frame/ethereum/src/lib.rs

```diff
@@ -290,7 +317,57 @@ pub mod pallet {
 				"pre log already exists; block is invalid",
 			);
 
-			Self::apply_validated_transaction(source, transaction)
+			Self::apply_validated_transaction(source, transaction, None)
+		}
+
+		/// Transact an Ethereum transaction with native WeightV2+ limit.
+		#[pallet::call_index(1)]
+		#[pallet::weight({ *weight_limit })]
+		pub fn transact_with_weight_limit(
+			origin: OriginFor<T>,
+			transaction: Transaction,
+			weight_limit: Weight,
+		) -> DispatchResultWithPostInfo {
+			let source = ensure_ethereum_transaction(origin)?;
+			// Disable transact functionality if PreLog exist.
+			assert!(
+				fp_consensus::find_pre_log(&frame_system::Pallet::<T>::digest()).is_err(),
+				"pre log already exists; block is invalid",
+			);
+
+			// Verify the transaction was signed with the same gas ratio as the native metrics in `weight_limit`
+			let weight_to_gas = U256::from(T::GasWeightMapping::weight_to_gas(
```

Comment:
We can't verify properly with the reverse way (recomputing gas from weight). We should instead recompute weight from the tx gas limit.
Also, it's better to move this check is `pre_dispatch_self_contained`

Reply:
resolved. this was part of the old dispatchable we were adding, weight_limit is computed now as part of the `fn transact` impl.